File: xabber/src/main/java/com/xabber/android/data/extension/references/mutable/filesharing/FileSharingExtension.java
package com.xabber.android.data.extension.references.mutable.filesharing;

import org.jivesoftware.smack.packet.ExtensionElement;
import org.jivesoftware.smack.util.XmlStringBuilder;

public class FileSharingExtension implements ExtensionElement {

    public static final String NAMESPACE = "https://xabber.com/protocol/files";
    public static final String FILE_SHARING_ELEMENT = "file-sharing";

    protected FileInfo fileInfo;
    protected FileSources fileSources;

    public FileSharingExtension(FileInfo fileInfo, FileSources fileSources) {
        this.fileInfo = fileInfo;
        this.fileSources = fileSources;
    }

    public void setFileInfo(FileInfo fileInfo) {
        this.fileInfo = fileInfo;
    }

    public FileInfo getFileInfo() {
        return fileInfo;
    }

    public void setFileSources(FileSources fileSources) {
        this.fileSources = fileSources;
    }

    public FileSources getFileSources() {
        return fileSources;
    }

    @Override
    public String getNamespace() {
        return NAMESPACE;
    }

    @Override
    public String getElementName() {
        return FILE_SHARING_ELEMENT;
    }

    public CharSequence toXML() {
        XmlStringBuilder xml = new XmlStringBuilder();
        xml.prelude(FILE_SHARING_ELEMENT, NAMESPACE);
        xml.rightAngleBracket();
        if (fileInfo != null) {
            xml.append(fileInfo.toXML());
        }
        if (fileSources != null) {
            xml.append(fileSources.toXML());
        }
        xml.closeElement(FILE_SHARING_ELEMENT);
        return xml;
    }

}

File: xabber/src/main/java/com/xabber/android/data/extension/references/mutable/filesharing/FileSharingExtension.kt
package com.xabber.android.data.extension.references.mutable.filesharing

import org.jivesoftware.smack.packet.ExtensionElement
import org.jivesoftware.smack.util.XmlStringBuilder

/**
 * Represents the <file-sharing/> element is used to represent information about a file according to [XEP-FILE]
 * This element has two allowable children: <file/> <sources/>
 *
 * @link https://xabber.com/protocol/otb
 */
class FileSharingExtension(var fileInfo: FileInfo, var fileSources: FileSources) : ExtensionElement {

    override fun getNamespace() = NAMESPACE

    override fun getElementName() = FILE_SHARING_ELEMENT

    override fun toXML() = XmlStringBuilder().apply {
        prelude(FILE_SHARING_ELEMENT, NAMESPACE)
        rightAngleBracket()
        append(fileInfo.toXML())
        append(fileSources.toXML())
        closeElement(FILE_SHARING_ELEMENT)
    }

    companion object {
        const val NAMESPACE = "https://xabber.com/protocol/files"
        const val FILE_SHARING_ELEMENT = "file-sharing"
    }

}